components/org.wso2.carbon.identity.oauth.ciba/src/main/java/org/wso2/carbon/identity/oauth/ciba/grant/CibaGrantHandler.java

```diff
@@ -0,0 +1,326 @@
+import org.wso2.carbon.identity.oauth2.IdentityOAuth2Exception;
+import org.wso2.carbon.identity.oauth2.dto.OAuth2AccessTokenRespDTO;
+import org.wso2.carbon.identity.oauth2.model.RequestParameter;
+import org.wso2.carbon.identity.oauth2.token.OAuthTokenReqMessageContext;
+import org.wso2.carbon.identity.oauth2.token.handlers.grant.AbstractAuthorizationGrantHandler;
+import org.wso2.carbon.identity.oauth2.util.OAuth2Util;
+
+import java.sql.Timestamp;
+import java.util.Calendar;
+import java.util.TimeZone;
+
+import static org.wso2.carbon.identity.oauth.ciba.common.CibaConstants.AUTH_REQ_ID;
+import static org.wso2.carbon.identity.oauth.ciba.common.CibaConstants.SEC_TO_MILLISEC_FACTOR;
+import static org.wso2.carbon.identity.oauth.ciba.exceptions.ErrorCodes.AUTHORIZATION_PENDING;
+import static org.wso2.carbon.identity.oauth.ciba.exceptions.ErrorCodes.EXPIRED_TOKEN;
+import static org.wso2.carbon.identity.oauth.ciba.exceptions.ErrorCodes.INVALID_REQUEST;
+import static org.wso2.carbon.identity.oauth.ciba.exceptions.ErrorCodes.SLOW_DOWN;
+import static org.wso2.carbon.identity.oauth.common.OAuth2ErrorCodes.OAuth2SubErrorCodes.CONSENT_DENIED;
+
+/**
+ * Grant Handler for CIBA.
+ */
+public class CibaGrantHandler extends AbstractAuthorizationGrantHandler {
+
+    // Used to keep the pre processed authorization code in the OAuthTokenReqMessageContext.
+    private static final String INVALID_GRANT = "invalid_grant";
+    private static final String MISSING_AUTH_REQ_ID = "auth_req_id_missing";
+    private static final String INVALID_AUTH_REQ_ID = "invalid auth_req_id";
+    private static final String INVALID_PARAMETERS = "invalid_request_parameters";
+    private static String cibaAuthCodeKey;
+
+    private static Log log = LogFactory.getLog(CibaGrantHandler.class);
+
+    @Override
+    public OAuth2AccessTokenRespDTO issue(OAuthTokenReqMessageContext tokReqMsgCtx) throws IdentityOAuth2Exception {
+
+        OAuth2AccessTokenRespDTO responseDTO = super.issue(tokReqMsgCtx);
+
+        try {
+            CibaDAOFactory.getInstance().getCibaAuthMgtDAO().updateStatus(cibaAuthCodeKey, AuthReqStatus.
+                    TOKEN_ISSUED);
+        } catch (CibaCoreException e) {
+            throw new IdentityOAuth2Exception("Error occurred in persisting status.", e);
+        }
+        return responseDTO;
+    }
+
+    @Override
+    public boolean validateGrant(OAuthTokenReqMessageContext tokReqMsgCtx) throws IdentityOAuth2Exception {
+
+        if (!super.validateGrant(tokReqMsgCtx)) {
+            return false;
+        }
+
+        // Obtain authentication request identifier from request.
+        String authReqId = getAuthReqId(tokReqMsgCtx);
+
+        try {
+            // Check whether provided authReqId is a valid.
+            validateAuthReqID(authReqId);
+
+            // Retrieving information from database and assign to CibaAuthCodeDO.
+            CibaAuthCodeDO cibaAuthCodeDO = retrieveCibaAuthCodeDO(authReqId);
```

Comment:
Can't we merge the two validations?

basically try to retrieve the DO from the authReqId. If a DO returns, it means authReqId is valid.